src/Parsers/ASTCreateFunctionQuery.cpp

```diff
@@ -50,4 +74,55 @@ String ASTCreateFunctionQuery::getFunctionName() const
     return name;
 }
 
+/// proton: starts
+Poco::JSON::Object::Ptr ASTCreateFunctionQuery::toJSON() const
+{
+    Poco::JSON::Object::Ptr func = new Poco::JSON::Object(Poco::JSON_PRESERVE_KEY_ORDER);
+    Poco::JSON::Object::Ptr inner_func = new Poco::JSON::Object(Poco::JSON_PRESERVE_KEY_ORDER);
+    inner_func->set("name", getFunctionName());
+    if (!is_javascript_func)
+    {
+        WriteBufferFromOwnString source_buf;
+        formatAST(*function_core, source_buf, false);
+        inner_func->set("source", source_buf.str());
+        inner_func->set("type", "sql");
+        func->set("function", inner_func);
+        return func;
+    }
+
+    assert(arguments && !arguments->children.empty());
+
+    Poco::JSON::Array json_args(Poco::JSON_PRESERVE_KEY_ORDER);
+    for (auto ast : arguments->children[0]->children)
+    {
+        Poco::JSON::Object::Ptr json_arg = new Poco::JSON::Object(Poco::JSON_PRESERVE_KEY_ORDER);
+        ASTNameTypePair * arg = ast->as<ASTNameTypePair>();
+        assert(arg);
+        json_arg->set("name", arg->name);
+        WriteBufferFromOwnString buf;
+        formatAST(*(arg->type), buf, false);
+        json_arg->set("type", buf.str());
+        json_args.add(json_arg);
+    }
+    inner_func->set("arguments", json_args);
+
+    /// type
+    inner_func->set("type", "javascript");
+
+    /// is_aggregation
+    inner_func->set("is_aggregation", is_aggregation);
+
+    /// return_type
+    WriteBufferFromOwnString return_buf;
+    formatAST(*return_type, return_buf, false);
+    inner_func->set("return_type", return_buf.str());
+
+    /// source
+    ASTJavaScriptFunction * js_func = function_core->as<ASTJavaScriptFunction>();
+    inner_func->set("source", js_func->source);
+
+    func->set("function", inner_func);
+    return func;
+}
```

Comment:
[notes]

```
{
  "function": {
    "arguments": [
      {"name": string, "type": string},
      ...
    ],
    "type": "javascript",
    "is_aggregation": boolean,
    "source": string,
    "return_type": string
  }
}
```

Reply:
[notes]
UserDefinedFunctionFactory registeFunction method only accept function definition in json format.